Update mod.rs

kiseitai3 opened the pull request:
Small clarification on the usage of `read_to_string` and `read_to_end` trait methods. The goal is to make it clear that these trait methods will become locked up if attempting to read to the end of stdin (which is a bit non-sensical unless the other end closes the pipe).

Fixes: rust-lang/rust#141714

cramertj committed:
Add Location::file_with_nul

This is useful for C/C++ APIs which expect the const char* returned
from __FILE__ or std::source_location::file_name.

joshtriplett committed:
Improve the documentation of `Display` and `FromStr`, and their interactions

In particular:
- `Display` is not necessarily lossless
- The output of `Display` might not be parseable by `FromStr`, and might
  not produce the same value if it is.
- Calling `.parse()` on the output of `Display` is usually a mistake
  unless a type's documented output and input formats match.
- The input formats accepted by `FromStr` depend on the type.

joshtriplett committed:
Add some more description of interactions between `Display` and `FromStr`

tgross35 committed:
Remove `i128` and `u128` from `improper_ctypes_definitions`

Rust's 128-bit integers have historically been incompatible with C [1].
However, there have been a number of changes in Rust and LLVM that
mean this is no longer the case:

* Incorrect alignment of `i128` on x86 [1]: adjusting Rust's alignment
  proposed at https://github.com/rust-lang/compiler-team/issues/683,
  implemented at https://github.com/rust-lang/rust/pull/116672.
* LLVM version of the above: resolved in LLVM, including ABI fix.
  Present in LLVM18 (our minimum supported version).
* Incorrect alignment of `i128` on 64-bit PowerPC, SPARC, and MIPS [2]:
  Rust's data layouts adjusted at
  https://github.com/rust-lang/rust/pull/132422,
  https://github.com/rust-lang/rust/pull/132741,
  https://github.com/rust-lang/rust/pull/134115.
* LLVM version of the above: done in LLVM 20
  https://github.com/llvm/llvm-project/issues/102783.
* Incorrect return convention of `i128` on Windows: adjusted to match
  GCC and Clang at https://github.com/rust-lang/rust/pull/134290.

At [3], the lang team considered it acceptable to remove `i128` from
`improper_ctypes_definitions` if the LLVM version is known to be
compatible. Time has elapsed since then and we have dropped support for
LLVM versions that do not have the x86 fixes, meaning a per-llvm-version
lint should no longer be necessary. The PowerPC, SPARC, and MIPS changes
only came in LLVM 20 but since Rust's datalayouts have also been updated
to match, we will be using the correct alignment regardless of LLVM
version.

`repr(i128)` was added to this lint in [4], but is also removed here.

Part of the decision is that `i128` should match `__int128` in C on
platforms that provide it, which documentation is updated to indicate.
We will not guarantee that `i128` matches `_BitInt(128)` since that can
be different from `__int128`. Some platforms (usually 32-bit) do not
provide `__int128`; if any ABIs are extended in the future to define it,
we will need to make sure that our ABI matches.

Closes: https://github.com/rust-lang/rust/issues/134288
Closes: https://github.com/rust-lang/rust/issues/128950

[1]: https://github.com/rust-lang/rust/issues/54341
[2]: https://github.com/rust-lang/rust/issues/128950
[3]: https://github.com/rust-lang/lang-team/issues/255#issuecomment-2088855084
[4]: https://github.com/rust-lang/rust/pull/138282

oli-obk committed:
Add `iter` macro

This adds an `iter!` macro that can be used to create movable
generators.

This also adds a yield_expr feature so the `yield` keyword can be used
within iter! macro bodies. This was needed because several unstable
features each need `yield` expressions, so this allows us to stabilize
them separately from any individual feature.

Co-authored-by: Oli Scherer <github35764891676564198441@oli-obk.de>
Co-authored-by: Jieyou Xu <jieyouxu@outlook.com>
Co-authored-by: Travis Cross <tc@traviscross.com>

psumbera committed:
build dist for x86_64-pc-solaris and sparcv9-sun-solaris

workingjubilee committed:
compiler: add CanonAbi

workingjubilee committed:
compiler: add AbiMap

- Add AbiMapping for encoding the nuance of deprecated ABIs

shepmaster committed:
Replace `elided_named_lifetimes` with `mismatched_lifetime_syntaxes`

tgross35 committed:
Run `builtins-test-intrinsics` when possible

Currently we only build this, but it is possible to run the binary.
Change the CI script to do so here.

shepmaster committed:
Introduce the `mismatched_lifetime_syntaxes` lint

tgross35 committed:
Replace the `nm` symbol check with a Rust implementation

This should be less error-prone and adaptable than the `nm` version, and
have better cross-platform support without needing LLVM `nm` installed.

tgross35 committed:
cleanup: Reuse `MinInt` and `Int` from `libm` in `compiler-builtins`

Since the two crates are now in the same repo, it is easier to share
code. Begin some deduplication with the integer traits.

tgross35 committed:
libm: Clean up unused files

These were deleted during refactoring in 0a2dc5d9 ("Combine the source
files for more generic implementations") but got added back by accident
in 54bac411 ("refactor: Move the libm crate to a subdirectory"). Remove
them again here.

amandasystems committed:
Move placeholder handling to a proper preprocessing step

This commit breaks out the logic of placheolder rewriting
into its own preprocessing step.

The only functional change from this is that the preprocessing
step (where extra `r: 'static` constraints are added) is
performed upstream of Polonius legacy, finally affecting
Polonius. That is mostly a by-product, though.

amandasystems committed:
Use an enum for SCC representatives, plus other code review

Co-authored-by: lcnr <rust@lcnr.de>

RalfJung committed:
UnsafePinned: also include the effects of UnsafeCell

RalfJung committed:
add Miri tests

shepmaster committed:
Rename `LifetimeSyntax` variants to lang-team-approved names

Lorl0rd committed:
fixed typo in readme

tgross35 committed:
Add tooling for `josh` syncs

Create a crate that handles pulling from and pushing to rust-lang/rust.
This can be invoked with the following:

    $ cargo run -p josh-sync -- rustc-pull
    $ RUSTC_GIT=/path/to/rust/checkout cargo run -p josh-sync -- rustc-push <username>

folkertdev committed:
add s390x z17 target features

nnethercote committed:
Remove `MetaItemParser::{path,deconstruct}`.

They're equivalent, and `path` is unused, and `deconstruct` has only one
call site outside of `path`.

nnethercote committed:
Remove `MetaItemListParser::all_{word,path}_list`, which are unused.

nnethercote committed:
Remove an unnecessary lifetime.

nnethercote committed:
Remove `MetaItemParser::{word,word_without_args,path_is}`.

They are unused.

nnethercote committed:
Rename `MetaItemParser::path_without_args` as `MetaItemParser::path`.

And avoid the clone.

RalfJung committed:
x86 (32/64): go back to passing SIMD vectors by-ptr

beetrees committed:
Enable `__powitf2` on MSVC

workingjubilee committed:
compiler: use CanonAbi for entry_abi

makes entry_abi a lowering of the ABI string, so now it can be
```json
  "entry_abi": "C",
  "entry_abi": "win64",
  "entry_abi": "aapcs",
```

workingjubilee committed:
compiler: change Conv to CanonAbi

workingjubilee committed:
cg_llvm: convert to CanonAbi

workingjubilee committed:
cg_clif: convert to CanonAbi

workingjubilee committed:
cg_gcc: convert to CanonAbi

workingjubilee committed:
miri: sed to CanonAbi

workingjubilee committed:
bless test for CanonAbi::X86(SysV64) change

joboet committed:
std: abort the process on failure to allocate a TLS key

The panic machinery uses TLS, so panicking if no TLS keys are left can lead to infinite recursion (see https://github.com/rust-lang/rust/issues/140798#issuecomment-2872307377). Rather than having separate logic for the panic count and the thread name, just always abort the process if a TLS key allocation fails. This also has the benefit of aligning the key-based TLS implementation with the documentation, which does not mention that a panic could also occur because of resource exhaustion.

tgross35 committed:
Update `CmpResult` to use a pointer-sized return type

As seen at [1], LLVM uses `long long` on LLP64 (to get a 64-bit integer
matching pointer size) and `long` on everything else, with exceptions
for AArch64 and AVR. Our current logic always uses an `i32`. This
happens to work because LLVM uses 32-bit instructions to check the
output on x86-64, but the GCC checks the full 64-bit register so garbage
in the upper half leads to incorrect results.

Update our return type to be `isize`, with exceptions for AArch64 and
AVR.

Fixes: https://github.com/rust-lang/compiler-builtins/issues/919

[1]: https://github.com/llvm/llvm-project/blob/0cf3c437c18ed27d9663d87804a9a15ff6874af2/compiler-rt/lib/builtins/fp_compare_impl.inc#L11-L27

cyrgani committed:
make `OsString::new` and `PathBuf::new` unstably const

workingjubilee committed:
compiler: simplify TargetOptions ABI functions

`adjust_abi` is not needed and `is_abi_supported` can be a 1-liner.

chenyukang committed:
Fix incorrect eq_unspanned in TokenStream

chenyukang committed:
remove eq_unspanned from TokenStream

oli-obk committed:
Use the informative error as the main const eval error message

tgross35 committed:
Add an empty rust-version file

This will be used by `josh` tooling.

tgross35 committed:
Remove unneeded C symbols

These are now provided by `compiler-builtins`, so there is no need to
also build the C versions. This was detected by checking for duplicate
symbols and not excluding weak symbols (like CI currently does).

D-Dario0 committed:
Typo in README.md

Link to Apache License changed from htps:// to https://

tgross35 committed:
Remove the now-unneeded llvm-tools-preview

Since a working `nm` is no longer needed as part of CI, the rustup
component can be removed.

tgross35 committed:
aarch64: Add a note saying why we use `frintx` rather than `frintn`

aDotInTheVoid committed:
jsondoclint: Extract `Command`/`CommandKind` into its own file

aDotInTheVoid committed:
jsondocck: command -> directive

tgross35 committed:
Reuse `libm`'s `Caat` and `CastFrom` in `compiler-builtins`

tgross35 committed:
Change `compiler-builtins` to edition 2024

Do the same for `builtins-test-intrinsics`. Mostly this means updating
`extern` to `unsafe extern`, and fixing a few new Clippy lints.

tgross35 committed:
symcheck: Print the command to make reproducing errors easier

tgross35 committed:
Add benchmarks for float parsing and printing

As part of this, the u256 benchmarks are reorganized to a group.

tgross35 committed:
Increase the benchmark rustc version to 2025-05-28

We may soon want to use some new nightly features in `compiler-builtins`
and `libm`, specifically `cfg_target_has_reliable_f16_f128` which was
added in the past few weeks. This will mean we need a newer toolchain
for benchmarks to continue building.

Bump to the current latest nightly so we are not blocked on this down
the line.

tgross35 committed:
ci: Allow concurrency outside of pull requests

When multiple merges to `master` happen before a CI run completes, the
in-progress job is getting canceled. Fix this by using the commit sha
for the group key if a pull request number is not available, rather than
`github.ref` (which is always `refs/head/master` after merge). This
should prevent jobs running on previous commits from getting cancelled,
while still ensuring there is only ever one active run per pull request.

tgross35 committed:
libm-test: Make `extensive` an attribute rather than a test type

Currently we run logspace tests for extensive tests, but there isn't any
reason we couldn't also run more kinds of tests more extensively (e.g.
more edge cases, combine edge cases with logspace for multi-input
functions, etc). As a first step toward making this possible, make
`extensive` a new field in `CheckCtx`, and rename `QuickSpaced` to
`Spaced`.

tgross35 committed:
ci: Allow for multiple icount benchmarks in the same run

We don't actually need this for now, but eventually it would be nice to
run icount benchmarks on multiple targets. Start tagging artifact names
with the architecture, and allow passing `--tag` to `ci-util.py` in
order to retrieve the correct one.

github-actions[bot] committed:
chore: release

oli-obk committed:
Clarify why we are talking about a failed const eval at a random place

Kobzol committed:
Do not run PGO/BOLT in x64 Linux alt builds

Signed-off-by: Jakub Beránek <berykubik@gmail.com>

workingjubilee committed:
compiler: actually remove Conv now that it is irrelevant

Kobzol committed:
Use 4-core large disk runner for the alt job

nia-e committed:
index: add method for checking range on DenseBitSet

compiler-errors committed:
Ensure query keys are printed with reduced queries

RalfJung committed:
abi_check: tweak some comments

tgross35 committed:
Upgrade all dependencies to the latest available version

In particular, this includes a fix to `iai-callgrind` that will allow us
to simplify our benchmark runner.

tgross35 committed:
Fix new `dead_code` warnings from recent nightlies

tgross35 committed:
ci: Refactor benchmark regression checks

iai-callgrind now correctly exits with error if regressions were found
[1], so we no longer need to check for regressions manually. Remove this
check and instead exit based on the exit status of the benchmark run.

[1] https://github.com/iai-callgrind/iai-callgrind/issues/337

Commit:
Use non-2015 edition paths in tests that do not test for their resolution

This allows for testing these tests on editions other than 2015

Derukugis committed:
Add link to correct documentation in htmldocck.py

Fix typo in src/etc/htmldocck.py

Co-authored-by: León Orell Valerian Liehr <me@fmease.dev>

Change documentation link to the correct section in src/etc/htmldocck.py

Co-authored-by: León Orell Valerian Liehr <me@fmease.dev>

Remove all top-level docs from htmldocck.py

usamoi committed:
remove f16: From<u16>

tshepang committed:
use a relative path, so that this also works offline

tshepang committed:
distracting indirection

smanilov committed:
Simplify long sentence

smanilov committed:
Trivial: dedup word

smanilov committed:
Add title and toc to Async chapter

This is standard for other chapters.

fmease committed:
Fix some warning blocks that contain Markdown

Contents inside of an HTML element only get interpreted as Markdown
(as opposed to HTML) if its separated from the HTML tags with line breaks.

tgross35 committed:
cleanup: Use `x.biteq(y)` rather than `x.to_bits() == y.to_bits()`

rs-sac committed:
Lightly tweak docs for BTree{Map,Set}::extract_if

- Move explanations into comments to match style
- Explain the second examples
- Make variable names match the data structure

cuviper committed:
Remove bootstrap cfgs from library/

tgross35 committed:
libm-test: Fix unintentional skips in `binop_common`

`binop_common` emits a `SKIP` that is intended to apply only to
`copysign`, but is instead applying to all binary operators. Correct the
general case but leave the currently-failing `maximum_num` tests as a
FIXME, to be resolved separately in [1].

Also simplify skip logic and NaN checking, and add a few more `copysign`
checks.

[1]: https://github.com/rust-lang/compiler-builtins/pull/939

azhogin committed:
Fix for async drop inside async gen fn

cuviper committed:
Remove more library bootstrap

kiseitai3 committed:
Update mod.rs

Small clarification on the usage of `read_to_string` and `read_to_end` trait methods.

kiseitai3 committed:
docs: Small clarification on the usage of `read_to_string` and `read_to_end` trait methods.

kiseitai3 committed:
docs: Small clarification on the usage of `read_to_string` and `read_to_end` trait methods.

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-19-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15404569469/job/43344691540) [(plain)](https://github.com/rust-lang/rust/commit/4d56e88cb8d825980d18820e054213b38bdde05f/checks/43344691540/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Diff in /checkout/library/std/src/io/mod.rs:919:
     ///
     /// # Notes
     ///
-    /// Be careful using this trait method with streams that are expected to be continuous. For example, using 
-    /// `read_to_end` with streams like `stdin` will simply lock the application into waiting on the 
+    /// Be careful using this trait method with streams that are expected to be continuous. For example, using
+    /// `read_to_end` with streams like `stdin` will simply lock the application into waiting on the
     /// transmission of data to conclude. It is recommended you use this method if you know the stream will be closed
     /// at the other end. The problem is that EOF or End of File is never reached for streams that never close or are finite.
     ///
Diff in /checkout/library/std/src/io/mod.rs:966:
     ///
     /// # Notes
     ///
-    /// Be careful using this trait method with streams that are expected to be continuous. For example, using 
-    /// `read_to_string` with streams like `stdin` will simply lock the application into waiting on the 
+    /// Be careful using this trait method with streams that are expected to be continuous. For example, using
+    /// `read_to_string` with streams like `stdin` will simply lock the application into waiting on the
     /// transmission of data to conclude. It is recommended you use this method if you know the stream will be closed
     /// at the other end. The problem is that EOF or End of File is never reached for streams that never close or are finite.
     ///
fmt: checked 6032 files
Build completed unsuccessfully in 0:00:41
  local time: Mon Jun  2 23:13:30 UTC 2025
  network time: Mon, 02 Jun 2025 23:13:31 GMT
```

</details>

nnethercote committed:
Remove pre-expansion AST stats.

They're very little value, because they only measure the top-level
`main.rs` or `lib.rs` file. (Other `.rs` files don't get read and parsed
until expansion occurs.)

I saw an example recently where the pre-expansion AST was 3KB in size
and the post-expansion AST was 66MB.

I kept the "POST EXPANSION" in the output header, I think that's useful
information to avoid possible confusion about when the measurement
happens.

Qelxiros committed:
add tests

nnethercote committed:
Remove `Path::is_ident`.

It checks that a path has a single segment that matches the given
symbol, and that there are zero generic arguments. It has a single use.

We also have `impl PartialEq<Symbol> for Path` which does exactly the
same thing *except* it doesn't check for zero generic arguments, which
seems like an oversight. It has numerous uses.

This commit removes `Path::is_ident`, adds a test for zero generic
arguments to `PartialEq<Symbol> for Path`, and changes the single use of
`is_ident` to instead use `==`.

tgross35 commented on library/std/src/io/mod.rs:
````diff
@@ -917,6 +917,13 @@ pub trait Read {
     /// # }
     /// ```
     ///
+    /// # Notes
+    ///
+    /// Be careful using this trait method with streams that are expected to be continuous. For example, using 
+    /// `read_to_end` with streams like `stdin` will simply lock the application into waiting on the 
+    /// transmission of data to conclude. It is recommended you use this method if you know the stream will be closed
+    /// at the other end. The problem is that EOF or End of File is never reached for streams that never close or are finite.
````
I think the wording could be simplified some. Perhaps something like this, with some cleanup:

> # Usage Notes
>
> `read_to_end` attempts to read a source until EOF, but many sources are continuous streams that do not send EOF; in these cases, `read_to_end` will block indefinitely. Standard input is one such stream which may be finite if piped, but is typically continuous without EOF when reading user input from a shell.
>
> Using `.lines()` with a [`BufReader`] can provide a better solution.

tgross35 commented on library/std/src/io/mod.rs:
````diff
@@ -957,6 +964,13 @@ pub trait Read {
     /// }
     /// ```
     ///
+    /// # Notes
+    ///
+    /// Be careful using this trait method with streams that are expected to be continuous. For example, using 
+    /// `read_to_string` with streams like `stdin` will simply lock the application into waiting on the 
+    /// transmission of data to conclude. It is recommended you use this method if you know the stream will be closed
+    /// at the other end. The problem is that EOF or End of File is never reached for streams that never close or are finite.
+    ///
     /// (See also the [`std::fs::read_to_string`] convenience function for
     /// reading from a file.)
````
Nit, could you move the new section under the existing "see also"? Since the "see also" is referring to reading a file in the example above.

tgross35 replied:
What are you running into? Usually fixing it looks something like this:

```
# If you haven't yet, add rust-lang/rust as a remote and update it
git remote add upstream git@github.com:rust-lang/rust.git
git fetch upstream master

# Do the rebase (from this branch)
git rebase -i upstream/master
```

Then for every line except the first replace `pick` with `f` (or `fixup`), and `git push --force-with-lease` to update this PR.

tgross35 reviewed (changes requested):
Thanks for the PR, I think this is a worthwhile change but the wording could use some adjustment.

Could you please also update the commit summary and PR description? They currently both say "Update mod.rs" but instead need to say what the actual change is.

Commit:
Add missing `dyn` keywords to tests that do not test for them Part 2

Commit:
Add missing 2015 edition directives

These tests specifically test 2015 edition behavior, so ensure that they can only be run with this edition

Commit:
Use non-2015 edition paths in tests that do not test for their resolution

This allows for testing these tests on editions other than 2015

jieyouxu committed:
Preparing for merge from rustc

Kivooeo committed:
cleaned up some tests

oli-obk committed:
Run wfcheck in one big loop instead of per module

Kivooeo committed:
cleaned up some tests

fmease committed:
Triagebot: Remove `assign.users_on_vacation`

onur-ozkan committed:
implement new `x` flag: `--skip-std-check-if-no-download-rustc`

Signed-off-by: onur-ozkan <work@onurozkan.dev>

onur-ozkan committed:
bless tidy

Signed-off-by: onur-ozkan <work@onurozkan.dev>

onur-ozkan committed:
add change-entry

Signed-off-by: onur-ozkan <work@onurozkan.dev>

compiler-errors committed:
Rework collect_and_apply to not rely on size hint for optimization

ZuseZ4 committed:
update gpu offload build command

compiler-errors committed:
Add a test to reproduce ICE

aDotInTheVoid committed:
rustdoc-json-type: Depend on `serde` and `serde_derive` seperately

Before this commit, serde_derive is built before serde. But serde does
not depend on serde_derive, so that is not needed. Instead, build serde
and serde_derive in parallel.

This speeds up compilation for users depending on rustdoc-json-types out
of tree.

Imports: https://www.github.com/rust-lang/rustdoc-types/pull/49
Co-authored-by: Martin Nordholts <martin.nordholts@codetale.se>

joshtriplett committed:
`FromStr`: Rework explanation of `FromStr`/`Display` round-tripping

- Drop the phrasing "usually a mistake".
- Mention that `Display` may not be lossless.
- Drop a misplaced parenthetical about round-tripping that didn't fit
  the paragraph it was in.

joshtriplett committed:
`Display`: Rework explanation of `FromStr`/`Display` round-tripping

- Drop "usually a mistake"
- Add phrasing from `FromStr` about round-tripping, and about how the
  inability to round-trip may surprise users.

lambdageek committed:
bootstrap: don't symlink source dir into stage0 sysroot

In StdLink::run we subsequently recursively copy the initial sysroot
lib directory into the stage0-sysroot lib directory.  If the initial
sysroot is a toolchain that includes the `rust-src` component (in
lib/rustlib/src/rust), if we add this symlink, that recursive copy
will overwrite the repo sources with the toolchain's sources.

Qelxiros committed:
add tests for negative numbers

Commit:
docs: Amended usage notes sections in read_to_end/read_to_string. Added usage note section to std::io::read_to_string as well.

Commit:
Merge remote-tracking branch 'origin/141714_stdin_read_to_string_docs' into 141714_stdin_read_to_string_docs

# Conflicts:
#	library/std/src/io/mod.rs

kiseitai3 committed:
Merge branch 'rust-lang:master' into 141714_stdin_read_to_string_docs

Commit:
Merge remote-tracking branch 'origin/141714_stdin_read_to_string_docs' into 141714_stdin_read_to_string_docs

rust-log-analyzer commented:
The job **`mingw-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15431196574/job/43429279647) [(plain)](https://github.com/rust-lang/rust/commit/ac993c4a682c6a8b9e15dd8eae8c77248c731052/checks/43429279647/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 Documenting std v0.0.0 (/checkout/library/std)
error: unresolved link to `read`
   --> library/std/src/io/mod.rs:929:58
    |
929 |     /// Using `.lines()` with a [`BufReader`] or using [`read`] can provide a better solution
    |                                                          ^^^^ no item named `read` in scope
    |
    = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
    = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(rustdoc::broken_intra_doc_links)]`

error: unresolved link to `read`
   --> library/std/src/io/mod.rs:983:58
    |
983 |     /// Using `.lines()` with a [`BufReader`] or using [`read`] can provide a better solution
    |                                                          ^^^^ no item named `read` in scope
    |
    = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`

error: unresolved link to `read`
    --> library/std/src/io/mod.rs:1297:54
     |
1297 | /// Using `.lines()` with a [`BufReader`] or using [`read`] can provide a better solution
     |                                                      ^^^^ no item named `read` in scope
     |
     = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`

error: could not document `std`
warning: build failed, waiting for other jobs to finish...
[RUSTC-TIMING] std test:false 7.749
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:01:01
  local time: Wed Jun  4 01:55:08 UTC 2025
  network time: Wed, 04 Jun 2025 01:55:08 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

tgross35 commented on library/std/src/io/mod.rs:
```diff
@@ -917,6 +917,17 @@ pub trait Read {
     ///
+    /// # Usage Notes
+    ///
+    /// `read_to_end` attempts to read a source until EOF, but many sources are continuous streams
+    /// that do not send EOF. In these cases, `read_to_end` will block indefinitely. Standard input
+    /// is one such stream which may be finite if piped, but is typically continuous. For example,
+    /// `cat <file> | <my_rust_program>` will correctly terminate with an `EOF` upon closure of cat.
+    /// `yes "Data" | pv | <my_rust_program` or `tail -f <file> | <my_rust_program>` may not close
+    /// the stream or insert an `EOF` termination character.
```
This looks much better to me. I think the `cat` example would be good to keep, but maybe translate the other example into words in case someone doesn't know what the `-f` flag does

```rust
    /// `read_to_end` attempts to read a source until EOF, but many sources are continuous streams
    /// that do not send EOF. In these cases, `read_to_end` will block indefinitely. Standard input
    /// is one such stream which may be finite if piped from a finite source (e.g. `cat foo.txt |
    /// my-program`), but is typically continuous (reading user input or sources that stay open).
```

nnethercote committed:
Visit the ident in `PreciseCapturingNonLifetimeArg`.

It's currently skipped, presumably by accident.

onur-ozkan committed:
run `check::Std` as the final step

Signed-off-by: onur-ozkan <work@onurozkan.dev>

onur-ozkan committed:
make it possible to request stage 0 std with `Builder::ensure`

Signed-off-by: onur-ozkan <work@onurozkan.dev>

onur-ozkan committed:
remove outdated RUSTC_SYSROOT handling for ci-rustc

Signed-off-by: onur-ozkan <work@onurozkan.dev>

nnethercote committed:
Improve some `Visitor` comments.

For AST/HIR/THIR visitors, explain the use of deconstruction.

scottmcm committed:
Change `tag_field` to `FieldIdx` in `Variants::Multiple`

It was already available as a generic parameter anyway, and it's not like we'll ever put a tag in the 5-billionth field.

oli-obk committed:
Report the actual item that evaluation failed for

Rageking8 committed:
Fix incorrect use of "recommend" over "recommended"

tgross35 committed:
Replace the musl submodule with a download script

The submodule was causing issues in rust-lang/rust, so eliminiate it
here. `build-musl` is also removed from `libm-test`'s default features
so the crate doesn't need to be built by default.

onur-ozkan committed:
document `skip-std-check-if-no-download-rustc` in rustc-dev-guide

Signed-off-by: onur-ozkan <work@onurozkan.dev>

onur-ozkan committed:
update `skip_std_check_if_no_download_rustc` doc-comments

Signed-off-by: onur-ozkan <work@onurozkan.dev>

m-ou-se committed:
Don't refer to 'this tail expression' in expansion.

The user has no clue what tail expression the compiler is talking
about: it is an implementation detail of the macro that it uses a block
with tail expression.

m-ou-se committed:
Don't refer to 'local binding' in extern macro.

The user has no clue what 'local binding' the compiler is talking about,
if they don't know the expansion of the macro.

smpdt committed:
bootstrap: Fix file permissions when dereferencing symlinks

matthewjasper committed:
Fix parsing of frontmatters with inner hyphens

tgross35 committed:
Update the upstream Rust version

To prepare for merging from rust-lang/rust, set the version file to:

    df8102fe5f Auto merge of #142002 - onur-ozkan:follow-ups2, r=jieyouxu

tgross35 committed:
`panic-handler`: Remove the `no_core` feature

This was introduced before `#[panic_handler]` was stable, but should no
longer be needed. Additionally, we only need it for
`builtins-test-intrinsics`, not as a dependency of `compiler-builtins`.

aDotInTheVoid committed:
jsondocck: Explain what `KNOWN_DIRECTIVE_NAMES` is doing

rust-log-analyzer commented:
The job **`mingw-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15480507221/job/43585436916) [(plain)](https://github.com/rust-lang/rust/commit/7298055ed4e3ff211effbbb39c51113eb88b44ad/checks/43585436916/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 Documenting std v0.0.0 (/checkout/library/std)
error: unresolved link to `read`
   --> library/std/src/io/mod.rs:929:58
    |
929 |     /// Using `.lines()` with a [`BufReader`] or using [`read`] can provide a better solution
    |                                                          ^^^^ no item named `read` in scope
    |
    = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
    = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(rustdoc::broken_intra_doc_links)]`

error: unresolved link to `read`
   --> library/std/src/io/mod.rs:983:58
    |
983 |     /// Using `.lines()` with a [`BufReader`] or using [`read`] can provide a better solution
    |                                                          ^^^^ no item named `read` in scope
    |
    = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`

error: unresolved link to `read`
    --> library/std/src/io/mod.rs:1297:54
     |
1297 | /// Using `.lines()` with a [`BufReader`] or using [`read`] can provide a better solution
     |                                                      ^^^^ no item named `read` in scope
     |
     = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`

error: could not document `std`
warning: build failed, waiting for other jobs to finish...
[RUSTC-TIMING] std test:false 7.970
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:01:03
  local time: Fri Jun  6 01:47:57 UTC 2025
  network time: Fri, 06 Jun 2025 01:47:57 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>
